tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,296 @@ static bool HasCanonicalDimensionNumbers(
+// the second_input_operand will get umap[{"c","b"}] -> (d2, d1)
+// the ouput_operand will get umap[{"a","c","b"}] -> (d0, d2, d1)
+SmallVector<AffineExpr> GetExprFromConfig(
+    const SmallVector<std::string>& loop_dims,
+    const DenseMap<StringRef, AffineExpr>& str_affine_dim_umap) {
+  SmallVector<AffineExpr> exprs;
+  for (const StringRef& dim : loop_dims) {
+    exprs.push_back(str_affine_dim_umap.lookup(dim));
+  }
+  return exprs;
+}
+
+// Convert the representation from string/vector<char> to vector<string>
+// i.e ("abc") -> {"a", "b", "c"}. For cases with ellipsis with batch rank 3:
+// get loop_dim = f("ab...cde") = {"a","b","0","1","2","c","d","e"}
+SmallVector<std::string> GetEinsumConfigAsVector(StringRef loop,
+                                                 const size_t& operand_rank) {
+  const std::string kEllipsis = "...";
+  SmallVector<std::string> loop_dim;
+  size_t pre_elip = loop.find(kEllipsis);
+  bool has_elip = pre_elip != std::string::npos;
+  if (!has_elip) pre_elip = loop.size();
+  // Add the dimension until the end or up to ellipsis if it exist
+  for (int pre_elip_ind = 0; pre_elip_ind < pre_elip; pre_elip_ind++)
+    loop_dim.push_back(std::string(1, loop[pre_elip_ind]));
+  if (!has_elip) return loop_dim;
+  // Case where Ellipsis presence:
+  size_t non_batch_rank = loop.size() - kEllipsis.size();
+  size_t batch_rank = operand_rank - non_batch_rank;
+  // Add the batch dimension ("0",...,"N") where N is rank of batch into the
+  // loop
+  for (int batch_ind = 0; batch_ind < batch_rank; batch_ind++)
+    loop_dim.push_back(std::to_string(batch_ind));
+  // Add the dimension after ellipsis into the loop
+  int post_elip = pre_elip + kEllipsis.size();
+  for (int post_elip_ind = post_elip; post_elip_ind < loop.size();
+       post_elip_ind++)
+    loop_dim.push_back(std::string(1, loop[post_elip_ind]));
+  return loop_dim;
+}
+
+// Check that all operand's batch has same rank
+bool CheckBatchHasEqualRank(const size_t& lhs_rank, StringRef lhs_loop,
+                            const size_t& rhs_rank, StringRef rhs_loop,
+                            const size_t& out_rank, StringRef out_loop) {
+  const std::string kEllipsis = "...";
+  SmallVector<int, 3> batch_rank_vec;
+  if (lhs_rank != lhs_loop.size()) {
+    size_t lhs_batch_rank = lhs_rank - (lhs_loop.size() - kEllipsis.size());
+    batch_rank_vec.push_back(lhs_batch_rank);
+  }
+  if (rhs_rank != rhs_loop.size()) {
+    size_t rhs_batch_rank = rhs_rank - (rhs_loop.size() - kEllipsis.size());
+    batch_rank_vec.push_back(rhs_batch_rank);
+  }
+  if (out_rank != out_loop.size()) {
+    size_t out_batch_rank = out_rank - (out_loop.size() - kEllipsis.size());
+    batch_rank_vec.push_back(out_batch_rank);
+  }
+  if (!std::equal(batch_rank_vec.begin() + 1, batch_rank_vec.end(),
+                  batch_rank_vec.begin()) &&
+      batch_rank_vec.size() > 1)
+    return false;
+  return true;
+}
+// Convert mhlo.einsum op into linalg.generic
```

Comment:
Add a blank line between comments and the above function.
```suggestion
}

// Convert mhlo.einsum op into linalg.generic
```